### PR TITLE
Replace usage of sentinel tripoint_min with cata::optional<>

### PR DESCRIPTION
**Pull request by BevapDin**
SUMMARY: Infrastructure "Replace usage of sentinel tripoint_min with cata::optional"

Changes variables/parameters/return types to be `cata::optional<tripoint>` instead of a plain `tripoint` and replace check for `tripoint_min` with check for the optional having a value.

### CODE REVIEW

File: src/debug_menu.cpp [hunk 1]
```diff
@@ -39,11 +39,12 @@ class mission_debug
 
 void teleport_short()
 {
-    const tripoint where( g->look_around() );
-    if( where == tripoint_min || where == g->u.pos() ) {
+const cata:
```
Comment (woutershep) [inline src/debug_menu.cpp]: This doesn't look right.

**Comment by AMurkin** [comment]
9a28da4
> (-1,+1) was returned twice

Actually 3 times. No (0,+1) too.

File: src/catalua.cpp [hunk 2]
```diff
@@ -386,6 +387,13 @@ class LuaValue
                 lua_setglobal( L, global_name );
             }
         }
+        static void push( lua_State *const L, const cata::opitional<T> &value ) {
```
Comment (AMurkin) [inline src/catalua.cpp]: Typo `cata::opitional`.

File: src/catalua.cpp [hunk 3]
```diff
@@ -494,6 +502,13 @@ class LuaReference : private LuaValue<T *>
             }
             LuaValue<T *>::push( L, value );
         }
+        static void push( lua_State *const L, const cata::opitional<T> &value ) {
```
Comment (AMurkin) [inline src/catalua.cpp]: Typo `cata::opitional`.

**Comment by Qrox** [comment]
```
In file included from src/string_formatter.h:6,
                 from src/debug.h:5,
                 from src/item.h:16,
                 from src/catalua.h:7,
                 from src/catalua.cpp:1:
src/optional.h: In instantiation of 'union cata::optional<Creature>::<unnamed>':
src/optional.h:34:9:   required from 'class cata::optional<Creature>'
src/catalua.cpp:914:61:   required from here
src/optional.h:36:24: error: cannot declare field 'cata::optional<Creature>::<unnamed union>::data' to be of abstract type 'Creature'
             StoredType data;
                        ^~~~
In file included from src/catalua.h:8,
                 from src/catalua.cpp:1:
src/creature.h:70:7: note:   because the following virtual functions are pure within 'Creature':
 class Creature
```

Looks like it's cause by this PR.

**Comment by ZhilkinSerg** [comment]
Huh, Travis and VS don't fail.

**Comment by KnightosMaster** [comment]
All commits after this was merged refuse to compile for me too.

File: src/ranged.cpp [hunk 4]
```diff
@@ -1200,9 +1200,11 @@ std::vector<tripoint> target_handler::target_ui( player &pc, target_mode mode,
         pc.cancel_activity();
 
         tripoint targ( 0, 0, 0 );
+        cata::optional<tripoint> mouse_pos;
         // Our coordinates will either be determined by coordinate input(mouse),
         // by a direction key, or by the previous value.
-        if( action == "SELECT" && ctxt.get_coordinates( g->w_terrain, targ.x, targ.y ) ) {
+        if( action == "SELECT" && ( mouse_pos = ctxt.get_coordinates( g->w_terrain ) ) ) {
```
Comment (ZhilkinSerg) [inline src/ranged.cpp]: Is it valid syntax? You have used result of assignment as condition. Will it return true only if assignment was successful?

Comment (AMurkin) [reply src/ranged.cpp]: `mouse_pos = ctxt.get_coordinates( g->w_terrain )` first assigns result of `ctxt.get_coordinates( g->w_terrain )` to `mouse_pos`, then evaluates to the value of `mouse_pos`.
Type was `bool` now `cata::optional<tripoint>` with `operator bool()` defined.

**Comment by BevapDin** [comment]
@KnightosMaster

> All commits after this was merged refuse to compile for me too.

Does the compiler tell you any reason, or is it just in a bad mood? How do you expect anyone to help you if you don't tell *what* happening? At least give use the error messages and the compiler version.

**Comment by Qrox** [comment]
I guess @KnightosMaster meant the same error I posted above, which is already fixed.

**Pull request merged**